### PR TITLE
Copter: Autotune Update

### PR DESCRIPTION
**Pull request by lthall**
This PR addresses a tuning issue when increasing angle P. It also reorders the testing functions for better readability.

### CODE REVIEW

File: ArduCopter/mode_autotune.cpp [hunk 1]
```diff
@@ -1182,36 +1184,89 @@ inline bool Copter::ModeAutoTune::yaw_enabled() {
...
+    // capture maximum rate
+    if (rate > meas_rate_max) {
+        // the measurement is continuing to increase without stopping
+        meas_rate_max = rate;
+        meas_rate_min = rate;
+    }
+
+    // capture maximum rate
+    if (rate < meas_rate_min) {
         // the measurement is continuing to increase without stopping
```
Comment (OXINARF) [inline ArduCopter/mode_autotune.cpp]: Both comments above seem wrong, as it's related to minimum.

Also, I don't know if it's on purpose, but I thought I should mention this check doesn't exist in `twitching_test_rate`.

File: ArduCopter/mode_autotune.cpp [hunk 2]
```diff
@@ -1332,44 +1396,36 @@ void Copter::ModeAutoTune::updating_d_down(float &tune_d, float tune_d_min, floa
...
         if (counter > 0 ) {
             counter--;
         }
-        // decrease P gain (which should decrease the maximum)
+        // decrease D gain (which should decrease bounce back)
+        tune_d -= tune_d*tune_d_step_ratio;
+        // stop tuning if we hit minimum D
+        if (tune_d <= tune_d_min) {
+            tune_d = tune_d_min;
+            Log_Write_Event(DATA_AUTOTUNE_REACHED_LIMIT);
```
Comment (OXINARF) [inline ArduCopter/mode_autotune.cpp]: I might be missing something, but how does this stop tuning? Looks like this is just moving code, but looking at other code places, I would say this needs to have `counter = AUTOTUNE_SUCCESS_COUNT;` or the comment removed.

File: ArduCopter/mode_autotune.cpp [hunk 3]
```diff
@@ -1332,44 +1396,36 @@ void Copter::ModeAutoTune::updating_d_down(float &tune_d, float tune_d_min, floa
...
+        }
+        // decrease P gain to match D gain reduction
         tune_p -= tune_p*tune_p_step_ratio;
-        // stop tuning if we hit maximum P
+        // stop tuning if we hit minimum P
         if (tune_p <= tune_p_min) {
             tune_p = tune_p_min;
-            counter = AUTOTUNE_SUCCESS_COUNT;
             Log_Write_Event(DATA_AUTOTUNE_REACHED_LIMIT);
```
Comment (OXINARF) [inline ArduCopter/mode_autotune.cpp]: Same.

File: ArduCopter/mode_autotune.cpp [hunk 4]
```diff
@@ -1332,44 +1396,36 @@ void Copter::ModeAutoTune::updating_d_down(float &tune_d, float tune_d_min, floa
...
-            ignore_next = false;
-        }
-    }else{
+    if (meas_rate_max > rate_target*(1+0.5f*g.autotune_aggressiveness)) {
         // ignore the next result unless it is the same as this one
         ignore_next = true;
-        // if maximum measurement was higher than target so decrement the success counter
+        // if maximum measurement was greater than target so increment the success counter
+        counter++;
+    } else if ((meas_rate_max < rate_target) && (meas_rate_max > rate_target*(1.0f-AUTOTUNE_D_UP_DOWN_MARGIN)) && (meas_rate_max-meas_rate_min > meas_rate_max*g.autotune_aggressiveness) && (tune_d > tune_d_min)) {
```
Comment (OXINARF) [inline ArduCopter/mode_autotune.cpp]: I think the last check should always be true (if tuning should indeed stop if `tune_d <= tune_d_min`)

File: ArduCopter/mode_autotune.cpp (hunk 1 above)
Comment (lthall) [reply ArduCopter/mode_autotune.cpp]: Thanks, I will correct them.

File: ArduCopter/mode_autotune.cpp (hunk 4 above)
Comment (lthall) [reply ArduCopter/mode_autotune.cpp]: yep, that is another comment problem I will fix it.

File: ArduCopter/mode_autotune.cpp (hunk 3 above)
Comment (lthall) [reply ArduCopter/mode_autotune.cpp]: yep

**Comment by rmackay9** [comment]
rebased and merged, thanks!

**Pull request closed without merging**